en/dev_log/ulog_file_format.md

```diff
@@ -41,233 +48,318 @@ The file consists of three sections:
-
-- 'I': information message.
-  ```c
-  struct message_info_s {
-  	struct message_header_s header;
-  	uint8_t key_len;
-  	char key[key_len];
-  	char value[header.msg_size-1-key_len]
-  };
-  ```
-  `key` is a plain string, as in the format message (can also be a custom type), but consists of only a single field without ending `;`, eg. `float[3] myvalues`.
-  `value` contains the data as described by `key`.
+:::note
+Message types described below are noted with the single alphabet `msg_type` characteristic in the beginning.
+:::
+
+
+### Definitions Section
+
+The definitions section contains basic information such as software version, uORB Topic format, initial parameter values, and so on.
+
+The message types in this section are:
+1. [Flag Bits](#b--flag-bits-message)
+2. [Format Definition](#f--format-definition-message)
+3. [Information](#i--information-message)
+4. [Multi Information](#m--multi-information-message)
+5. [Parameter](#p--parameter-message)
+6. [Default Parameter](#q--default-parameter-message)
+
+
+#### 'B': Flag Bits Message
+
+:::note
+This message must be the **first message** right after the header section, so that it has a fixed constant offset from the start of the file!
+:::
+
+```
+struct ulog_message_flag_bits_s {
+  struct message_header_s;
+  uint8_t compat_flags[8];
+  uint8_t incompat_flags[8];
+  uint64_t appended_offsets[3]; ///< file offset(s) for appended data if appending bit is set
+};
+```
+
+- `compat_flags`: compatible flag bits.
+  - `compat_flags[0]`, bit 0, *DEFAULT_PARAMETERS*: if set, the log contains parameter defaults (message 'Q').
+
+  The rest of the bits is currently not defined and all must be set to 0.
+  These bits can be used for future ULog changes that are compatible with existing parsers.
+  It means parsers can just ignore the bits if one of the unknown bits is set.
+- `incompat_flags`: incompatible flag bits.
+  The LSB bit of index 0 is set to one if the log contains appended data and at least one of the `appended_offsets` is non-zero.
+  All other bits are undefined and must be set to 0.
+  If a parser finds one of these bits set, it must refuse to parse the log.
+  This can be used to introduce breaking changes that existing parsers cannot handle.
+- `appended_offsets`: File offsets (0-based) for appended data.
+  If no data is appended, all offsets must be zero.
+  This can be used to reliably append data for logs that may stop in the middle of a message.
+
+  A process appending data should do:
+  - set the relevant `incompat_flags` bit,
+  - set the first `appended_offsets` that is 0 to the length of the log file,
+  - then append any type of messages that are valid for the Data section.
+
+It is possible that there are more fields appended at the end of this message in future ULog specifications.
+This means a parser must not assume a fixed length of this message.
+If the message is longer than expected (currently 40 bytes), the exceeding bytes must just be ignored.
```

Comment:
This part is still very vague / hard to understand

Reply:
Yes, that is odd wording. @bkueng Is this what is meant? 
```suggestion
It is possible that there are more fields appended at the end of this message in future ULog specifications.
This means a parser must not assume a fixed length of this message.
If the message is longer than expected (currently 40 bytes), any additional bytes must be ignored/discarded.
```

Reply:
Yes